zanata-war/src/main/java/org/zanata/service/impl/TranslationUpdatedManager.java

```diff
@@ -55,23 +61,30 @@ public void textFlowStateUpdated(
     // Fire asynchronous event
     void publishAsyncEvent(TextFlowTargetStateEvent event) {
         if (BeanManagerProvider.isActive()) {
-            int wordCount = textFlowDAO.getWordCount(event.getTextFlowId());
+            Long versionId = event.getKey().getProjectIterationId();
+            Long documentId = event.getKey().getDocumentId();
+            LocaleId localeId = event.getKey().getLocaleId();
 
-            // TODO use Event.fire()
-            BeanManagerProvider.getInstance().getBeanManager().fireEvent(
-                    new DocumentStatisticUpdatedEvent(
-                            event.getProjectIterationId(),
-                            event.getDocumentId(), event.getLocaleId(),
-                            wordCount,
-                            event.getPreviousState(), event.getNewState())
-            );
+            for(TextFlowTargetState state: event.getStates()) {
+                int wordCount =
+                    textFlowDAO.getWordCount(state.getTextFlowId());
+                // TODO: turn DocumentStatisticUpdatedEvent into an aggregate event like TextFlowTargetStateEvent
+                // TODO: generate one DocumentStatisticUpdatedEvent per TextFlowTargetStateEvent
```

Comment:
Sorry, can you make that TODO PERF? It looks like DocumentServiceImpl.documentStatisticUpdated is the only observer for DocumentStatisticUpdatedEvent, and it would really benefit from batching, especially for projects which have webhooks enabled.

// TODO PERF: generate one DocumentStatisticUpdatedEvent per TextFlowTargetStateEvent. See DocumentServiceImpl.documentStatisticUpdated